[4.x only] fix(router): do not finish bootstrap until all the routes are resolved

vsavkin opened the pull request:
Updated 2.x branch https://github.com/angular/angular/pull/14327

vicb commented on modules/@angular/platform-webworker/src/web_workers/worker/location_providers.ts:
```diff
@@ -24,9 +24,14 @@ export const WORKER_APP_LOCATION_PROVIDERS = [
+  {provide: LOCATION_INITIALIZED, useFactory: locationInitialized, deps: [PlatformLocation]}
 ];
 
+function locationInitialized(platformLocation: WebWorkerPlatformLocation) {
```
should it be exported ?

vsavkin replied:
You are right. It should be. As far as I know webworkers do not work with AoT.

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -320,12 +332,24 @@ export class Router {
   navigated: boolean = false;
 
   /**
+   * Used by RouterModule and experimental libraries. This allows us to
```
can we remove the reference to "experimental libs" here (I doubt we will update the comment later on) ?

vsavkin replied:
Done

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -679,17 +703,23 @@ export class Router {
         urlAndSnapshot$ = of ({appliedUrl: url, snapshot: precreatedState});
       }
 
+      const beforePreactivationDone$ = mergeMap.call(urlAndSnapshot$, (p: any) => {
```
any ->type ?

vsavkin replied:
Done

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -679,17 +703,23 @@ export class Router {
-        return {appliedUrl, snapshot};
-      });
+      const preactivationTraverse$ =
+          map.call(beforePreactivationDone$, ({appliedUrl, snapshot}: any) => {
```
any ->type

vsavkin replied:
Done

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -708,11 +738,20 @@ export class Router {
         }
       });
 
+      const preactivationDone$ = mergeMap.call(preactivationResolveData$, (p: any) => {
```
any -> type

vsavkin replied:
Done

vicb replied:
?

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -320,12 +332,24 @@ export class Router {
+   * @internal
+   */
+  hooks: {beforePreactivation: RouterHook, afterPreactivation: RouterHook} = {
+    beforePreactivation: defaultRouterHook,
```
should the default be null ? (what is the benefit of having an explicit default ?)

vsavkin replied:
so we don't have to check for null and use polymorphism to handle this case. We use the same approach to handle other defaults in the router (e.g., defaultErrorHandler)

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -278,22 +280,77 @@ export function rootRoute(router: Router): ActivatedRoute {
+  constructor(private injector: Injector) {}
+
+  appInitializer(): Promise<any> {
+    const p: Promise<any> = this.injector.get(LOCATION_INITIALIZED, new Promise(res => res()));
```
`new Promise(res => res())` -> `Promise.resolve(null)` ?

vsavkin replied:
Done

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -278,22 +280,77 @@ export function rootRoute(router: Router): ActivatedRoute {
+  appInitializer(): Promise<any> {
+    const p: Promise<any> = this.injector.get(LOCATION_INITIALIZED, new Promise(res => res()));
+    return p.then(() => {
+      let resolve: any = null;
```
Function ?

vsavkin replied:
Done

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -278,6 +278,18 @@ type NavigationParams = {
+ * @internal
+ */
+function defaultRouterHook(snapshot: RouterStateSnapshot): Observable<RouterStateSnapshot> {
+  return of (snapshot);
```
do we have testsfor the return value ?

vsavkin replied:
Every integration test checks this value since it is used by the navigation pipeline. If the value were incorrect, the navigation results would have been different.

vicb replied:
makes sense

vicb commented:
Thanks for the PR and sorry for the review delay.

Mostly nits, missing tests ?

vsavkin commented:
Thanks for reviewing it! I addressed the comments. 

I created bootstrap.spec.ts that tests the behaviour. What do you think should be tested separately?

vsavkin commented:
Btw, I also have a version of this PR for the 2.x branch #14327

vicb commented on modules/@angular/router/src/router.ts:
```diff
@@ -278,6 +278,18 @@ type NavigationParams = {
+ * @internal
+ */
+function defaultRouterHook(snapshot: RouterStateSnapshot): Observable<void> {
+  return of (null);
```
void vs null ?

vsavkin replied:
`void` isn't an expression. I can use `undefined`, but I don't think we use undefined very much in the codebase, that's why I chose null.

vicb commented on modules/@angular/router/src/router_module.ts:
```diff
@@ -278,22 +280,77 @@ export function rootRoute(router: Router): ActivatedRoute {
+ *
+ * We need to start the navigation in a APP_INITIALIZER to block the bootstrap if
+ * a resolver or a guards executes asynchronously. Second, we need to actually run
+ * activation in a BOOTSTRAP_LISTENER. We utilize an experimental afterPreactivation
```
remove "experimental"

vsavkin replied:
Fixed

mhevery commented:
@vicb this breaks g3. Could you have a look?

vsavkin commented:
Update: I cherry-picked the commit onto the 2.x branch. You can find it here: https://github.com/angular/angular/pull/14327

vsavkin committed:
fix(router): do not finish bootstrap until all the routes are resolved

Fixes #12162
